day8/model.go

```diff
@@ -0,0 +1,200 @@
+package day8
+
+import (
+	"fmt"
+	"strings"
+)
+
+type SignalPattern string
+
+func parseSignalPatterns(s string) ([]SignalPattern, error) {
+	signals := strings.Fields(s)
+	if len(signals) != 10 {
+		return nil, fmt.Errorf("incorrect number of signals in %v", s)
+	}
+	results := make([]SignalPattern, len(signals))
+	for i, s := range signals {
+		results[i] = SignalPattern(s)
+	}
+	return results, nil
+}
+
+func (s SignalPattern) Single() segment {
+	return segment(s[0])
+}
+
+func (s SignalPattern) WithoutPattern(other ...SignalPattern) SignalPattern {
+	var b strings.Builder
+OUTER:
+	for _, r := range s {
+		for _, o := range other {
+			if strings.ContainsRune(string(o), r) {
+				continue OUTER
+			}
+		}
+		// must not have been in any of our "others"
+		b.WriteRune(r)
+
+	}
+	return SignalPattern(b.String())
+}
+
+type SegmentPattern string
+
+func parseSegmentPatterns(s string) ([]SegmentPattern, error) {
+	segments := strings.Fields(s)
+	if len(segments) != 4 {
+		return nil, fmt.Errorf("incorrect number of segments in %v", s)
+	}
+	results := make([]SegmentPattern, len(segments))
+	for i, s := range segments {
+		results[i] = SegmentPattern(s)
+	}
+	return results, nil
+}
+
+type segment rune
```

Comment:
I think some of this might have been easier if I was consistent about this being the base type; `string`, `[]rune`, `[]segment` and `SignalPattern` are all used to mean kinda the same thing. Centralizing on `[]segment` and _giving that a type name_ might have cleaned some of this up.